[HWToBTOR2] operator not are not currently supported

Hi, when I try to generate btor2 from a divider example in Chisel, it reports an error:
`variadic operations not are not currently supported`

more details: 
```
home/dhp/work/chisel-prove/examples/src/main/scala/examples/Radix2Divider.scala:72:31: error: variadic operations not are not currently supported
home/dhp/work/chisel-prove/examples/src/main/scala/examples/Radix2Divider.scala:72:31: note: see current operation: %212 = "comb.concat"(%13, %58, %211) : (i1, i1, i5) -> i7
```

related code snippet:
```scala
    val canSkipShift = (len.U + Log2(bReg)) - Log2(aValx2Reg)
```

--- a/lib/Conversion/HWToBTOR2/HWToBTOR2.cpp
+++ b/lib/Conversion/HWToBTOR2/HWToBTOR2.cpp
@@ -274,12 +274,6 @@ private:
   // and a result width.
   void genBinOp(StringRef inst, Operation *binop, Value op1, Value op2,
                 size_t width) {
-    if (binop->getNumOperands() != 2) {
-      binop->emitError(
-          "only the binary form of this operation is currently supported");
-      return;
-    }
-
     // Set the LID for this operation
     size_t opLID = getOpLID(binop);
 
@@ -301,7 +295,11 @@ private:
     auto operands = op->getOperands();
     size_t sid = sortToLIDMap.at(width);
 
-    // No need to consider 0-operand case as it'll be rejected by the verifier
+    if (operands.size() == 0) {
+      op->emitError("variadic operations with no operands are not supported");
+      return;
+    }
+
     // If there's only one operand, then we don't generate a BTOR2 instruction,
     // we just reuse the operand's existing LID
     if (operands.size() == 1) {
@@ -313,25 +311,31 @@ private:
       return;
     }
 
-    // TODO: this excludes concat, which requires special handling for result
-    // sorts
-    if (!op->hasTrait<mlir::OpTrait::SameOperandsAndResultType>()) {
-      op->emitError(
-          "variadic operations with differing operand and result types are "
-          "not currently supported");
-      return;
-    }
+    // Special case for concat since intermediate results need different sorts
+    auto isConcat = isa<comb::ConcatOp>(op);
 
     // Unroll variadic op into series of binary ops
     // This will represent the previous operand in the chain:
     auto prevOperandLID = getOpLID(operands[0]);
 
+    // Track the current width so we can work out new types if this is a concat
+    auto currentWidth = operands[0].getType().getIntOrFloatBitWidth();
+
     for (auto operand : operands.drop_front()) {
       // Manually increment lid since we need multiple per op
+
+      if (isConcat) {
+        // For concat, the sort width increases with each operand
+        currentWidth += operand.getType().getIntOrFloatBitWidth();
+        // Ensure that the sort exists
+        genSort("bitvec", currentWidth);
+      }
+
       auto thisLid = lid++;
       auto thisOperandLID = getOpLID(operand);
-      os << thisLid << " " << inst << " " << sid << " " << prevOperandLID << " "
-         << thisOperandLID << "\n";
+      os << thisLid << " " << inst << " "
+         << (isConcat ? sortToLIDMap.at(currentWidth) : sid) << " "
+         << prevOperandLID << " " << thisOperandLID << "\n";
       prevOperandLID = thisLid;
     }
 
@@ -734,7 +738,7 @@ public:
   void visitComb(comb::AndOp op) { visitVariadicOp(op, "and"); }
   void visitComb(comb::OrOp op) { visitVariadicOp(op, "or"); }
   void visitComb(comb::XorOp op) { visitVariadicOp(op, "xor"); }
-  void visitComb(comb::ConcatOp op) { visitBinOp(op, "concat"); }
+  void visitComb(comb::ConcatOp op) { visitVariadicOp(op, "concat"); }
 
   // Extract ops translate to a slice operation in btor2 in a one-to-one
   // manner

--- a/test/Conversion/HWToBTOR2/comb.mlir
+++ b/test/Conversion/HWToBTOR2/comb.mlir
@@ -60,6 +60,12 @@ module {
     %11 = comb.and %10 : i32
     %12 = comb.and %a, %11 : i32
 
+    // CHECK: [[NID20:[0-9]+]] sort bitvec 96
+    // CHECK: [[NID21:[0-9]+]] sort bitvec 64
+    // CHECK: [[NID22:[0-9]+]] concat [[NID21]] 2 2
+    // CHECK: [[NID23:[0-9]+]] concat [[NID20]] [[NID22]] 2
+    %13 = comb.concat %a, %a, %a : i32, i32, i32
+
     // CHECK:   [[ASSERTNID1:[0-9]+]] implies [[NID3]] [[NID5]] [[NID12]]
     // CHECK:   [[ASSERTNID2:[0-9]+]] not [[NID3]] [[ASSERTNID1]]
     // CHECK:   [[ASSERTNID3:[0-9]+]] bad [[ASSERTNID2:[0-9]+]]

--- a/test/Conversion/HWToBTOR2/errors.mlir
+++ b/test/Conversion/HWToBTOR2/errors.mlir
@@ -26,7 +26,7 @@ hw.module @reg_with_argument_initial(in %clk: !seq.clock, in %in: i32, in %init:
 
 // -----
 
-hw.module @variadic_op(in %input: i32) {
-  // expected-error @below {{only the binary form of this operation is currently supported}}
-  %and = comb.concat %input, %input, %input : i32, i32, i32
+hw.module @nullary_variadic() {
+  // expected-error @below {{variadic operations with no operands are not supported}}
+  "comb.concat"() : () -> (i0)
 }

